src/ai/backend/manager/repositories/permission_controller/role_manager.py

```diff
@@ -163,13 +165,113 @@ async def add_object_permission_to_user_role(
+            .options(selectinload(RoleRow.permission_group_rows))
+        )
+        role_row = cast(Optional[RoleRow], role_row)
+        if role_row is None:
+            raise UserSystemRoleNotFound(f"System role for user {user_id} not found")
+        return role_row
+
+    async def _query_role_by_id(
+        self,
+        db_session: SASession,
+        role_id: uuid.UUID,
+    ) -> RoleDataWithPermissions:
+        role_row = await db_session.scalar(
+            sa.select(RoleRow)
+            .where(RoleRow.id == role_id)
+            .options(
+                selectinload(RoleRow.permission_group_rows).options(
+                    selectinload(PermissionGroupRow.permission_rows)
+                ),
+                selectinload(RoleRow.object_permission_rows),
+            )
+        )
+        role_row = cast(Optional[RoleRow], role_row)
+        if role_row is None:
+            raise RoleNotFound(f"Role with id {role_id} not found")
+        return role_row
+
+    async def _query_entity_associated_scopes(
+        self,
+        db_session: SASession,
+        entity_id: ObjectId,
+    ) -> list[ScopeId]:
+        raw_scope_entity_rows = await db_session.scalars(
+            sa.select(AssociationScopesEntitiesRow).where(
+                sa.and_(
+                    AssociationScopesEntitiesRow.entity_type == entity_id.entity_type,
+                    AssociationScopesEntitiesRow.entity_id == entity_id.entity_id,
+                )
+            )
+        )
+        scope_entity_rows = cast(list[AssociationScopesEntitiesRow], raw_scope_entity_rows.all())
+        return [
+            ScopeId(scope_type=row.scope_type, scope_id=row.scope_id) for row in scope_entity_rows
+        ]
+
+    async def _add_permission_groups_to_role_if_not_exist(
+        self,
+        db_session: SASession,
+        role_row: RoleRow,
+        scope_ids: Collection[ScopeId],
+    ) -> None:
+        scope_ids_to_add = {scope for scope in scope_ids}
+        for permission_group in role_row.permission_group_rows:
+            scope_id = permission_group.parsed_scope_id()
+            scope_ids_to_add.discard(scope_id)
+        creators = [
+            PermissionGroupCreator(
+                role_id=role_row.id,
+                scope_id=scope_id,
+            )
+            for scope_id in scope_ids_to_add
+        ]
+
+        rows = [PermissionGroupRow.from_input(creator) for creator in creators]
+        db_session.add_all(rows)
+        await db_session.flush()
+
+    async def _add_object_permissions_to_role(
+        self,
+        db_session: SASession,
+        role_id: uuid.UUID,
+        entity_id: ObjectId,
+        operations: Collection[OperationType],
+    ) -> None:
+        if not operations:
+            log.warning(
+                "no operations provided to add object permissions to role {}, skipping.", role_id
+            )
+            return
```

Comment:
This is a log that should not actually occur in properly written code, so if this log occurs, it should be fixed immediately.

Reply:
Should an error be raised here?